client/app/reducers/ManageAvailabilityReducer.js

```diff
@@ -1,5 +1,135 @@
+import moment from 'moment';
+import { isSameDay } from 'react-dates';
+import * as actionTypes from '../constants/ManageAvailabilityConstants';
 
-export default function manageAvailabilityReducer(state = initialState) {
-  return state;
-}
+const initialState = new Map({
+  isOpen: true,
+  visibleMonth: moment().startOf('month'),
+
+  // List of days that buyers have reserved. These cannot be blocked.
+  reservedDays: new List(),
+
+  // List of days that are blocked and save to the API.
+  blockedDays: new List(),
+
+  // List of changes with `action` (String) and `day` (moment
+  // instance) keys. Whenever the user blocks/allows a day, a change
+  // is added to this list. The actual changes for the UI and the API
+  // are calculated by going through this list to see the final values
+  // for each day and comparing those to the blocked days above.
+  changes: new List(),
+});
+
+export const EDIT_VIEW_OPEN_HASH = 'manage-availability';
+
+const includesDay = (days, day) =>
+      days.some((d) => isSameDay(d, day));
+
+const ACTION_ALLOW = 'allow';
+const ACTION_BLOCK = 'block';
+
+const withChange = (state, action, day) => {
+  if (includesDay(state.get('reservedDays'), day)) {
+    // Changes to reserved days ignored
+    return state;
+  }
+  const change = new Map({ action, day });
+  return state.set('changes', state.get('changes').push(change));
+};
+
+// Calculate all unique changes to the original blocked days
+const compressedChanges = (state) => {
+  const changes = state.get('changes');
+
+  const isSameDayChange = (c1) => (c2) =>
+        isSameDay(c1.get('day'), c2.get('day'));
+
+  // Compress to only take the final state of a single day into account
+  const compressed = changes.reduce((result, change) => {
+    const existingChange = result.find(isSameDayChange(change));
+
+    if (existingChange && existingChange.get('action') !== change.get('action')) {
+      return result.filterNot(isSameDayChange(change)).push(change);
+    } else if (!existingChange) {
+      return result.push(change);
+    }
+    return result;
+  }, new List());
+
+  const blockedDays = state.get('blockedDays');
+
+  // Only include changes to the original blocked days
+  return compressed.filter((c) => {
+    const isBlock = c.get('action') === ACTION_BLOCK;
+    const alreadyBlocked = includesDay(blockedDays, c.get('day'));
+
+    // A change is only considered if it is a block that isn't already
+    // blocked or if it is an allow to a day that is already blocked.
+    return (isBlock && !alreadyBlocked) || (!isBlock && alreadyBlocked);
+  });
+};
+
+export const hasChanges = (state) =>
+  compressedChanges(state).size > 0;
+
+// Calculate currently blocked days from the fetched ones and the
+// unsaved changes.
+export const blockedDays = (state) => {
+  const changes = compressedChanges(state);
+
+  // Collect lists of day instances for allows and blocks
+  const splitChanges = changes.reduce((result, c) => {
```

Comment:
Ah, now I remember why I didn't use `groupBy`: you'd need another iteration to "pluck" only the days from the change instances.